src/bootstrap/src/core/build_steps/compile.rs

```diff
@@ -136,7 +142,8 @@ impl Step for Std {
             compiler: run.builder.compiler(run.builder.top_stage, run.build_triple()),
             target: run.target,
             crates,
-            force_recompile: false,
+            // Force compiling from source as it was explicitly called with `x build library (or std/sysroot)`.
+            force_recompile: !run.builder.paths.is_empty(),
```

Comment:
It seems like this will be recompilation even if e.g. `x.py build compiler` is passed? I'm not sure that gating on just any path is the right call. Shouldn't this use something like "is std modified?" perhaps?

Reply:
> It seems like this will be recompilation even if e.g. x.py build compiler is passed? I'm not sure that gating on just any path is the right call.

No, that shouldn't happen.

> Shouldn't this use something like "is std modified?" perhaps?

We can do that but it's not necessary in my opinion. If user explicitly asks for library build, we can do it right away without checking if "is std modified?".

Reply:
I think it's pretty confusing -- certainly against my expectations -- that the same source tree can produce *different* results depending on the command (x.py build vs. x.py build library, for example). I'd expect that to always yield the same results, though the path-less (`x.py build`) would perhaps build more.

If we choose whether to rebuild std based on the paths provided, that violates this basic assumption that I think we have so far upheld.

Reply:
[Changed](https://github.com/rust-lang/rust/pull/127974/commits/bf7fb649a24320cd78758944f0274f8483c7478d) into `if modified` logic.